src/hotspot/share/nmt/virtualMemoryTracker.cpp

```diff
@@ -47,545 +47,210 @@ void VirtualMemory::update_peak(size_t size) {
-    rgn->exclude_region(addr, exclude_size);
-
-    // higher part
-    address high_base = addr + size;
-    size_t  high_size = top - high_base;
-
-    CommittedMemoryRegion high_rgn(high_base, high_size, *rgn->call_stack());
-    LinkedListNode<CommittedMemoryRegion>* high_node = _committed_regions.add(high_rgn);
-    assert(high_node == nullptr || node->next() == high_node, "Should be right after");
-    return (high_node != nullptr);
-  }
-
-  return false;
+void VirtualMemoryTracker::set_reserved_region_tag(address addr, MemTag mem_tag) {
+    ReservedMemoryRegion rgn = tree()->find_reserved_region(addr);
+    VMATree::SummaryDiff diff = tree()->set_tag((VMATree::position) addr, rgn.size(), mem_tag);
+    apply_summary_diff(diff);
 }
 
-bool ReservedMemoryRegion::remove_uncommitted_region(address addr, size_t sz) {
-  assert(addr != nullptr, "Invalid address");
-  assert(sz > 0, "Invalid size");
-
-  CommittedMemoryRegion del_rgn(addr, sz, *call_stack());
-  address end = addr + sz;
-
-  LinkedListNode<CommittedMemoryRegion>* head = _committed_regions.head();
-  LinkedListNode<CommittedMemoryRegion>* prev = nullptr;
-  CommittedMemoryRegion* crgn;
-
-  while (head != nullptr) {
-    crgn = head->data();
-
-    if (crgn->same_region(addr, sz)) {
-      VirtualMemorySummary::record_uncommitted_memory(crgn->size(), mem_tag());
-      _committed_regions.remove_after(prev);
-      return true;
-    }
-
-    // del_rgn contains crgn
-    if (del_rgn.contain_region(crgn->base(), crgn->size())) {
-      VirtualMemorySummary::record_uncommitted_memory(crgn->size(), mem_tag());
-      head = head->next();
-      _committed_regions.remove_after(prev);
-      continue;  // don't update head or prev
-    }
-
-    // Found addr in the current crgn. There are 2 subcases:
-    if (crgn->contain_address(addr)) {
+void VirtualMemoryTracker::Instance::apply_summary_diff(VMATree::SummaryDiff diff) {
+  assert(_tracker != nullptr, "Sanity check");
+  _tracker->apply_summary_diff(diff);
+}
 
-      // (1) Found addr+size in current crgn as well. (del_rgn is contained in crgn)
-      if (crgn->contain_address(end - 1)) {
-        VirtualMemorySummary::record_uncommitted_memory(sz, mem_tag());
-        return remove_uncommitted_region(head, addr, sz); // done!
+void VirtualMemoryTracker::apply_summary_diff(VMATree::SummaryDiff diff) {
+  VMATree::SingleDiff::delta reserve_delta, commit_delta;
+  size_t reserved, committed;
+  MemTag tag = mtNone;
+  auto print_err = [&](const char* str) {
+    return;
+    // Will be commented out after 8350567. The error outputs corrupt the jdk-image.
+    // log_error(nmt)("summary mismatch, at %s, for %s,"
+    //                " diff-reserved:  %ld"
+    //                " diff-committed: %ld"
+    //                " vms-reserved: %zu"
+    //                " vms-committed: %zu",
+    //                str, NMTUtil::tag_to_name(tag), (long)reserve_delta, (long)commit_delta, reserved, committed);
+  };
+
```

Comment:
8350567 is merged now! I think that that PR should be merged in.

Reply:
Under test.